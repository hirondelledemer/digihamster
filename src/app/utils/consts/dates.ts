export const DAY = 1000 * 60 * 60 * 24;
export const WEEK = DAY * 7;
