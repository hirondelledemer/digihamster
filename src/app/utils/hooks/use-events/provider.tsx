"use client";
import { ReactNode, useCallback, useEffect, useReducer } from "react";
import { useToast } from "@/app/components/ui/use-toast";

import { reducer } from "./reducer";

import { Event } from "@/models/event";

import { EventsStateAction, EventsStateActionType } from "./actions";
import { api, FieldsRequired } from "./api";
import { EventsStateContext } from "./state-context";
import { EventsActionsContext } from "./actions-context";

const handleApiError = (
  error: any,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  const errorMessage =
    error.response?.data?.message || "An unexpected error occurred";
  toast({
    title: "Error",
    description: errorMessage,
    variant: "destructive",
  });
};

const handleSuccessToast = (
  toast: ReturnType<typeof useToast>["toast"],
  message: string
) => {
  toast({
    title: "Success",
    description: message,
  });
};

const fetchEvents = async (
  dispatch: React.Dispatch<EventsStateAction>,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  try {
    dispatch({ type: EventsStateActionType.StartLoading });
    const eventsResponse = await api.getEvents();

    dispatch({
      type: EventsStateActionType.FinishLoading,
      payload: { data: eventsResponse.data },
    });
  } catch (err) {
    dispatch({
      type: EventsStateActionType.Error,
      payload: { errorMessage: err },
    });
    handleApiError(err, toast);
  }
};

export const EventsContextProvider = ({
  children,
}: {
  children: ReactNode;
}) => {
  const [state, dispatch] = useReducer(reducer, {
    isLoading: false,
    data: [],
  });

  const { toast } = useToast();

  const fetchEventsMemoized = useCallback(() => {
    fetchEvents(dispatch, toast);
  }, [toast]);

  useEffect(() => {
    fetchEventsMemoized();
  }, [fetchEventsMemoized]);

  const createEvent = useCallback(
    async (data: FieldsRequired, onDone?: () => void) => {
      const tempId = "temp-id";

      const tempEvent: Event = {
        _id: tempId,
        completed: false,
        deleted: false,
        createdAt: "",
        updatedAt: "",
        tags: [],
        ...data,
      };
      dispatch({
        type: EventsStateActionType.CreateEvent,
        payload: { event: tempEvent },
      });

      if (onDone) {
        onDone();
      }

      try {
        const response = await api.createEvent(data);

        dispatch({
          type: EventsStateActionType.UpdateEvent,
          payload: {
            id: tempId,
            event: response.data,
          },
        });

        handleSuccessToast(toast, "Event has been created");
      } catch (e: any) {
        // todo: fix
        dispatch({
          type: EventsStateActionType.DeleteEvent,
          payload: {
            id: tempId,
          },
        });
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  const updateEvent = useCallback(
    async (eventId: string, props: Partial<Event>, onDone?: () => void) => {
      try {
        dispatch({
          type: EventsStateActionType.UpdateEvent,
          payload: {
            id: eventId,
            event: props,
          },
        });
        if (onDone) {
          onDone();
        }

        await api.updateEvent(eventId, props);

        handleSuccessToast(toast, "Event has been updated");
      } catch (e: any) {
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  const deleteEvent = useCallback(
    async (eventId: string, onDone?: () => void) => {
      try {
        dispatch({
          type: EventsStateActionType.DeleteEvent,
          payload: {
            id: eventId,
          },
        });
        if (onDone) {
          onDone();
        }

        await api.deleteEvent(eventId);

        handleSuccessToast(toast, "Event has been deleted");
      } catch (e: any) {
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  return (
    <EventsStateContext.Provider value={state}>
      <EventsActionsContext.Provider
        value={{
          create: createEvent,
          update: updateEvent,
          delete: deleteEvent,
        }}
      >
        {children}
      </EventsActionsContext.Provider>
    </EventsStateContext.Provider>
  );
};
