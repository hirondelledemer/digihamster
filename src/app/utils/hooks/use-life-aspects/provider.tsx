"use client";
import { ReactNode, useCallback, useEffect, useReducer } from "react";
import { useToast } from "@/app/components/ui/use-toast";

import { reducer } from "./reducer";

import { LifeAspectStateAction, LifeAspectStateActionType } from "./actions";
import { api } from "./api";
import { LifeAspectsStateContext } from "./state-context";

const handleApiError = (
  error: any,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  const errorMessage =
    error.response?.data?.message || "An unexpected error occurred";
  toast({
    title: "Error",
    description: errorMessage,
    variant: "destructive",
  });
};

const fetchLifeAspects = async (
  dispatch: React.Dispatch<LifeAspectStateAction>,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  try {
    dispatch({ type: LifeAspectStateActionType.StartLoading });
    const response = await api.getLifeAspects();

    dispatch({
      type: LifeAspectStateActionType.FinishLoading,
      payload: { data: response.data },
    });
  } catch (err) {
    dispatch({
      type: LifeAspectStateActionType.Error,
      payload: { errorMessage: err },
    });
    handleApiError(err, toast);
  }
};

export const LifeAspectsContextProvider = ({
  children,
}: {
  children: ReactNode;
}) => {
  const [state, dispatch] = useReducer(reducer, {
    isLoading: false,
    data: [],
  });

  const { toast } = useToast();

  const fetchDataMemoized = useCallback(() => {
    fetchLifeAspects(dispatch, toast);
  }, [toast]);

  useEffect(() => {
    fetchDataMemoized();
  }, [fetchDataMemoized]);

  return (
    <LifeAspectsStateContext.Provider value={state}>
      {children}
    </LifeAspectsStateContext.Provider>
  );
};
