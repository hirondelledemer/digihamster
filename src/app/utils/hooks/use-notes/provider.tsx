"use client";
import { ReactNode, useCallback, useEffect, useReducer } from "react";
import { useToast } from "@/app/components/ui/use-toast";

import { reducer } from "./reducer";

import { NotesStateAction, NotesStateActionType } from "./actions";
import { api, CreateNoteParams } from "./api";
import { NotesStateContext } from "./state-context";
import { NotesActionsContext } from "./actions-context";
import { Note } from "@/models/note";

const handleApiError = (
  error: any,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  const errorMessage =
    error.response?.data?.message || "An unexpected error occurred";
  toast({
    title: "Error",
    description: errorMessage,
    variant: "destructive",
  });
};

const handleSuccessToast = (
  toast: ReturnType<typeof useToast>["toast"],
  message: string
) => {
  toast({
    title: "Success",
    description: message,
  });
};

const fetchNotes = async (
  dispatch: React.Dispatch<NotesStateAction>,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  try {
    dispatch({ type: NotesStateActionType.StartLoading });
    const response = await api.getNotes();

    dispatch({
      type: NotesStateActionType.FinishLoading,
      payload: { data: response.data },
    });
  } catch (err) {
    dispatch({
      type: NotesStateActionType.Error,
      payload: { errorMessage: err },
    });
    handleApiError(err, toast);
  }
};

export const NotesContextProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(reducer, {
    isLoading: false,
    data: [],
  });

  const { toast } = useToast();

  const fetchDataMemoized = useCallback(() => {
    fetchNotes(dispatch, toast);
  }, [toast]);

  useEffect(() => {
    fetchDataMemoized();
  }, [fetchDataMemoized]);

  const createNote = useCallback(
    async (data: CreateNoteParams, onDone?: () => void) => {
      const tempId = "temp-id";

      const tempNote: Note = {
        _id: tempId,
        deleted: false,
        createdAt: "",
        updatedAt: "",
        isActive: false,
        userId: "", // todo: why this is not required in other places?
        ...data,
      };
      dispatch({
        type: NotesStateActionType.CreateNote,
        payload: { note: tempNote },
      });

      if (onDone) {
        onDone();
      }

      try {
        const response = await api.createNote(data);

        dispatch({
          type: NotesStateActionType.UpdateNote,
          payload: {
            id: tempId,
            note: response.data,
          },
        });

        handleSuccessToast(toast, "Note has been created");
      } catch (e: any) {
        // todo: fix
        dispatch({
          type: NotesStateActionType.DeleteNote,
          payload: {
            id: tempId,
          },
        });
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  const updateNote = useCallback(
    async (id: string, props: Partial<Note>, onDone?: () => void) => {
      try {
        dispatch({
          type: NotesStateActionType.UpdateNote,
          payload: {
            id,
            note: props,
          },
        });
        if (onDone) {
          onDone();
        }

        await api.updateNote(id, props);

        handleSuccessToast(toast, "Note has been updated");
      } catch (e: any) {
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  const deleteNote = useCallback(
    async (id: string, onDone?: () => void) => {
      try {
        dispatch({
          type: NotesStateActionType.DeleteNote,
          payload: {
            id: id,
          },
        });
        if (onDone) {
          onDone();
        }

        await api.deleteNote(id);

        handleSuccessToast(toast, "Note has been deleted");
      } catch (e: any) {
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  return (
    <NotesStateContext.Provider value={state}>
      <NotesActionsContext.Provider
        value={{
          create: createNote,
          update: updateNote,
          delete: deleteNote,
        }}
      >
        {children}
      </NotesActionsContext.Provider>
    </NotesStateContext.Provider>
  );
};
