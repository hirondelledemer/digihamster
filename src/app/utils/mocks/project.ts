import { Project } from "@/models/project";

export const generateProject: (
  i?: number,
  properties?: Partial<Project>
) => Project = (i = 1, properties) => {
  return {
    _id: `project${i}`,
    title: `Project ${i}`,
    color: `project-color-${i}`,
    order: 0,
    deleted: false,
    createdAt: 0,
    updatedAt: 0,
    ...properties,
  };
};

export const generateListOfProjects: (count: number) => Project[] = (count) => {
  return [...Array(count)].map((_v, i) => generateProject(i));
};

export const generateCustomProjectsList: (
  taskInfo: Partial<Project>[]
) => Project[] = (projectInfo) => {
  return projectInfo.map((projectProperties, i) => ({
    ...generateProject(i, projectProperties),
  }));
};
