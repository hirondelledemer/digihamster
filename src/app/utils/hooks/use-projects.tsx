"use client";

import {
  Dispatch,
  SetStateAction,
  createContext,
  useContext,
  useEffect,
  useState,
} from "react";
import axios from "axios";
import { Project } from "@/models/project";
import { useToast } from "@/app/components/ui/use-toast";
import { updateObjById } from "../common/update-array";
import { arrayMove } from "@dnd-kit/sortable";

type FieldsRequired = "title" | "color" | "disabled";

export interface ProjectsContextValue {
  data: Project[];
  defaultProject?: Project;
  setData: Dispatch<SetStateAction<Project[]>>;
  error?: unknown;
  loading: boolean;
  updateProject(id: string, props: Partial<Project>, onDone?: () => void): void;
  updateProjectsOrder(
    movedProjectId: string,
    overProjectId: string,
    onDone?: () => void
  ): void;
  createProject(data: Pick<Project, FieldsRequired>): void;
  getProjectById(id: string): Project;
}

export const ProjectsContext = createContext<ProjectsContextValue>({
  data: [],
  setData: () => {},
  loading: false,
  updateProject: () => {},
  createProject: () => {},
  updateProjectsOrder: () => {},
  getProjectById: () => {
    return null as unknown as Project;
  },
});

const { Provider } = ProjectsContext;

export const ProjectsContextProvider = ({ children }: any) => {
  const [data, setData] = useState<Project[]>([]);
  const [defaultProject, setDefaultProject] = useState<Project>();
  const [error, setError] = useState<unknown>(null);
  const [loading, setLoading] = useState(false);
  const { toast } = useToast();

  useEffect(() => {
    (async function () {
      try {
        setLoading(true);
        const projectsResponse = await axios.get<{
          projects: Project[];
          defaultProject: Project;
        }>("/api/projects");
        setData(projectsResponse.data.projects);
        setDefaultProject(projectsResponse.data.defaultProject);
      } catch (err) {
        setError(err);
        toast({
          title: "Error",
          description: "error while getting projects",
          variant: "destructive",
        });
      } finally {
        setLoading(false);
      }
    })();
  }, [toast]);

  const createProject = async (data: Pick<Project, FieldsRequired>) => {
    const tempId = "temp-id";

    const tempProject: Project = {
      _id: tempId,
      deleted: false,
      order: 0,
      ...data,
    };
    setData((e) => [...e, tempProject]);

    try {
      const response = await axios.post<Project>("/api/projects", data);
      setData((e) => updateObjById<Project>(e, tempId, response.data));
      toast({
        title: "Success",
        description: "Project has been created",
      });
    } catch (e) {
      toast({
        title: "Error",
        description: JSON.stringify(e),
        variant: "destructive",
      });
    }
  };

  const updateProject = async (
    projectId: string,
    props: Partial<Project>,
    onDone?: () => void
  ) => {
    try {
      setData((p) =>
        updateObjById<Project>(p, projectId, {
          ...props,
        })
      );
      onDone && onDone();
      await axios.patch("/api/projects", {
        id: projectId,
        ...props,
      });
      toast({
        title: "Success",
        description: "Project has been updated",
      });
    } catch (e) {
      toast({
        title: "Error",
        description: JSON.stringify(e),
        variant: "destructive",
      });
    }
  };

  const updateProjectsOrder = async (
    movedProjectId: string,
    overProjectId: string,
    onDone?: () => void
  ) => {
    try {
      const oldIndex = data.findIndex((p) => p._id === movedProjectId);
      const newIndex = data.findIndex((p) => p._id === overProjectId);
      const newArray = arrayMove(data, oldIndex, newIndex);

      setData(newArray);
      onDone && onDone();
      await axios.patch("/api/projects/sort", {
        sortOrder: newArray.map((p, index) => ({
          projectId: p._id,
          order: index,
        })),
      });
      toast({
        title: "Success",
        description: "Project order been updated",
      });
    } catch (e) {
      toast({
        title: "Error",
        description: JSON.stringify(e),
        variant: "destructive",
      });
    }
  };

  const getProjectById = (id: string) => {
    const project = data.find((project) => project._id === id);
    if (!project) {
      throw Error("no project by this id");
    }

    return project;
  };

  return (
    <Provider
      value={{
        data,
        setData,
        error,
        loading,
        defaultProject,
        updateProject,
        createProject,
        updateProjectsOrder,
        getProjectById,
      }}
    >
      {children}
    </Provider>
  );
};

export default function useProjects() {
  const {
    data,
    setData,
    defaultProject,
    updateProject,
    createProject,
    updateProjectsOrder,
    getProjectById,
    loading,
  } = useContext(ProjectsContext);

  return {
    data,
    setData,
    defaultProject,
    updateProject,
    createProject,
    updateProjectsOrder,
    getProjectById,
    loading,
  };
}
