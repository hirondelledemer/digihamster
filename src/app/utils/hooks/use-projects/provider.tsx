"use client";
import { ReactNode, useCallback, useEffect, useReducer } from "react";
import { useToast } from "@/app/components/ui/use-toast";

import { reducer } from "./reducer";

import { ProjectsStateAction, ProjectsStateActionType } from "./actions";
import { api, FieldsRequired } from "./api";
import { ProjectsStateContext } from "./state-context";
import { ProjectsActionsContext } from "./actions-context";
import { Project } from "@/models/project";

const handleApiError = (
  error: any,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  const errorMessage =
    error.response?.data?.message || "An unexpected error occurred";
  toast({
    title: "Error",
    description: errorMessage,
    variant: "destructive",
  });
};

const handleSuccessToast = (
  toast: ReturnType<typeof useToast>["toast"],
  message: string
) => {
  toast({
    title: "Success",
    description: message,
  });
};

const fetchProjects = async (
  dispatch: React.Dispatch<ProjectsStateAction>,
  toast: ReturnType<typeof useToast>["toast"]
) => {
  try {
    dispatch({ type: ProjectsStateActionType.StartLoading });
    const projectResponse = await api.getProjects();

    dispatch({
      type: ProjectsStateActionType.FinishLoading,
      payload: {
        data: projectResponse.data.projects,
        defaultProject: projectResponse.data.defaultProject,
      },
    });
  } catch (err) {
    dispatch({
      type: ProjectsStateActionType.Error,
      payload: { errorMessage: err },
    });
    handleApiError(err, toast);
  }
};

export const ProjectsContextProvider = ({
  children,
}: {
  children: ReactNode;
}) => {
  const [state, dispatch] = useReducer(reducer, {
    isLoading: true,
    data: [] as Project[],
    defaultProject: null,
  });

  const { toast } = useToast();

  const fetchProjectsMemoized = useCallback(() => {
    fetchProjects(dispatch, toast);
  }, [toast]);

  useEffect(() => {
    fetchProjectsMemoized();
  }, [fetchProjectsMemoized]);

  const createProject = useCallback(
    async (data: FieldsRequired, onDone?: () => void) => {
      const tempId = "temp-id";

      const tempProject: Project = {
        _id: tempId,
        deleted: false,
        order: 0,
        ...data,
      };
      dispatch({
        type: ProjectsStateActionType.CreateProject,
        payload: { project: tempProject },
      });

      if (onDone) {
        onDone();
      }

      try {
        const response = await api.createProject(data);

        dispatch({
          type: ProjectsStateActionType.UpdateProject,
          payload: {
            id: tempId,
            project: response.data,
          },
        });

        handleSuccessToast(toast, "Project has been created");
      } catch (e: any) {
        // todo: fix
        dispatch({
          type: ProjectsStateActionType.DeleteProject,
          payload: {
            id: tempId,
          },
        });
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  const updateProject = useCallback(
    async (projectId: string, props: Partial<Project>, onDone?: () => void) => {
      try {
        dispatch({
          type: ProjectsStateActionType.UpdateProject,
          payload: {
            id: projectId,
            project: props,
          },
        });
        if (onDone) {
          onDone();
        }

        await api.updateProject(projectId, props);

        handleSuccessToast(toast, "Project has been updated");
      } catch (e: any) {
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  const deleteProject = useCallback(
    async (projectId: string, onDone?: () => void) => {
      try {
        dispatch({
          type: ProjectsStateActionType.DeleteProject,
          payload: {
            id: projectId,
          },
        });
        if (onDone) {
          onDone();
        }

        await api.deleteProject(projectId);

        handleSuccessToast(toast, "Project has been deleted");
      } catch (e: any) {
        const errorMessage =
          e.response?.data?.message || "An unexpected error occurred";
        handleApiError(errorMessage, toast);
      }
    },
    [toast]
  );

  const updateOrder = async (
    movedProjectId: string,
    overProjectId: string,
    onDone?: () => void
  ) => {
    try {
      dispatch({
        type: ProjectsStateActionType.UpdateOrder,
        payload: {
          movedProjectId,
          overProjectId,
        },
      });

      if (onDone) {
        onDone();
      }

      await api.updateOrder({
        sortOrder: state.data.map((p, index) => ({
          projectId: p._id,
          order: index,
        })),
      });
      handleSuccessToast(toast, "Project order been updated");
    } catch (e: any) {
      const errorMessage =
        e.response?.data?.message || "An unexpected error occurred";
      handleApiError(errorMessage, toast);
    }
  };

  const getProjectById = (id: string) => {
    return state.data.find((project) => project._id === id) || null;
  };

  return (
    <ProjectsStateContext.Provider value={{ ...state, getProjectById }}>
      <ProjectsActionsContext.Provider
        value={{
          create: createProject,
          update: updateProject,
          delete: deleteProject,
          updateOrder,
        }}
      >
        {children}
      </ProjectsActionsContext.Provider>
    </ProjectsStateContext.Provider>
  );
};
