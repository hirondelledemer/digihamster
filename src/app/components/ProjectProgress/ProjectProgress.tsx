"use client";
import { now } from "@/app/utils/date/date";
import { useProjectsState } from "@/app/utils/hooks/use-projects/state-context";
import useTasks from "@/app/utils/hooks/use-tasks";
import { Project } from "@/models/project";
import { addDays, differenceInCalendarDays, format, isValid } from "date-fns";
import { useRouter } from "next/navigation";

import React, { FC, useCallback, useMemo } from "react";

export interface ProjectProgressProps {
  testId?: string;
}

const ProjectProgressBar: FC<{ project: Project }> = ({
  project,
}): JSX.Element => {
  const { data: tasks } = useTasks();

  const router = useRouter();

  const allTasks = useMemo(
    () => tasks.filter((task) => task.projectId === project._id),
    [project._id, tasks]
  );

  const activeTaskCount = useMemo(
    () => allTasks.filter((task) => task.isActive && !task.completed).length,
    [allTasks]
  );

  const getCompletionEta = useCallback(() => {
    const completedTasksDates = allTasks
      .filter((task) => task.completed)
      .sort((a, b) => a.completedAt! - b.completedAt!)
      .map((task) => task.completedAt);

    const remainingTasksCount = allTasks.filter(
      (task) => !task.completed
    ).length;

    const dateDiff =
      differenceInCalendarDays(now(), completedTasksDates[0]!) || 1;

    const completionRate = completedTasksDates.length / dateDiff;

    const etaDate = addDays(now(), remainingTasksCount * completionRate);

    if (isValid(etaDate)) {
      return format(etaDate, "MMM dd");
    }
    return "date unknown";
  }, [allTasks]);

  const completedTaskCount = useMemo(
    () => allTasks.filter((task) => task.completed).length,
    [allTasks]
  );

  return (
    <div
      key={project._id}
      className="flex flex-col gap-5 hover:font-semibold cursor-pointer"
      style={{ color: project.color }}
      onClick={() => router.push(`/?projectId=${project._id}`, undefined)}
    >
      <div className="flex flex-col">
        <div className="text-sm">
          {project.title} ({getCompletionEta()})
        </div>
        <div
          data-testid="progress-bar"
          className="w-full h-2"
          style={{
            border: `1px solid ${project.color}`,
            borderRadius: "3px",
          }}
        >
          <div
            data-testid="progress-bar-outer"
            style={{
              height: "100%",
              width: `${
                ((activeTaskCount + completedTaskCount) / allTasks.length) * 100
              }%`,
              background: `repeating-linear-gradient(135deg, ${project.color}, ${project.color} 2px, transparent 2px, transparent 4px)`,
            }}
          >
            <div
              data-testid="progress-bar-inner"
              style={{
                height: "100%",
                backgroundColor: project.color,

                width: `${
                  (completedTaskCount /
                    (activeTaskCount + completedTaskCount)) *
                  100
                }%`,
              }}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

const ProjectProgress: FC<ProjectProgressProps> = (): JSX.Element => {
  const { data: projects } = useProjectsState();

  const activeProjects = projects.filter((project) => !project.disabled);

  return (
    <div className="flex flex-col gap-3">
      {activeProjects.map((project) => (
        <ProjectProgressBar key={project._id} project={project} />
      ))}
    </div>
  );
};

export default ProjectProgress;
