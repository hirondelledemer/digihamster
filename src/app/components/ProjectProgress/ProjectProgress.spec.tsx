import { render, screen } from "@/config/utils/test-utils";
import ProjectProgress from "./ProjectProgress";
import mockAxios from "jest-mock-axios";
import { wrapWithTasksProvider } from "@/app/utils/tests/wraps";
import { generateCustomTasksList } from "@/app/utils/mocks/task";
import { DAY } from "@/app/utils/consts/dates";
import { generateProject } from "@/app/utils/mocks/project";
import { ProjectsContextProvider } from "@/app/utils/hooks/use-projects/provider";

jest.mock("../../utils/date/date");

describe("ProgressProject", () => {
  const PROJECT = generateProject();

  const renderComponent = () =>
    render(
      wrapWithTasksProvider(
        <ProjectsContextProvider>
          <ProjectProgress />
        </ProjectsContextProvider>,
        {
          data: generateCustomTasksList([
            {
              completed: true,
              completedAt: 0 - DAY,
              projectId: PROJECT._id,
            },
            {
              completed: true,
              completedAt: 0 - DAY,
              projectId: PROJECT._id,
            },
            {
              completed: true,
              completedAt: 0 - DAY * 3,
              projectId: PROJECT._id,
            },
            {
              completed: false,
              isActive: true,
              projectId: PROJECT._id,
            },
            {
              completed: false,
              projectId: PROJECT._id,
            },
            {
              completed: false,
              projectId: PROJECT._id,
            },
          ]),
        }
      )
    );

  afterEach(() => {
    mockAxios.reset();
  });

  it("should render DailyProgress", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: { projects: [PROJECT] } });

    renderComponent();
    const bar = await screen.findByTestId("progress-bar");

    expect(bar).toMatchSnapshot();
  });
});
