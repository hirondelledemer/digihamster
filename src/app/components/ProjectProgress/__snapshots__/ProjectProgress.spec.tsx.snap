// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressProject should render DailyProgress 1`] = `
<div
  class="w-full h-2"
  data-testid="progress-bar"
  style="border-radius: 3px;"
>
  <div
    data-testid="progress-bar-outer"
    style="height: 100%; width: 66.66666666666666%;"
  >
    <div
      data-testid="progress-bar-inner"
      style="height: 100%; width: 25%;"
    />
  </div>
</div>
`;
